File: hanoi-towers.js
/**
 * @description solving the Hanoi Towers puzzle
 * @param numDiscs
 */


function solveHanoiTowersLoop(numDiscs){
  var recurrenceSolution = 0;
  for (var i = 0; i < numDiscs; i++){
    recurrenceSolution = 2 * recurrenceSolution + 1;
  }
  return recurrenceSolution;
} //O(n)

function solveHanoiTowersRecurrently(numDiscs){
  if (numDiscs == 0) return 0;
  return 2*solveHanoiTowersRecurrently(numDiscs-1) + 1;
} //O(n)

function solveHanoiTowersClosedForm(numDiscs){
  return Math.pow(2, numDiscs) - 1;
} //O(1)

File: hanoi-towers.swift
import Foundation

func solveHanoiTowersLoop(numDiscs:UInt) -> UInt {
  var recurrenceSolution: UInt = 0
  for _ in 0..<numDiscs {
    recurrenceSolution = 2 * recurrenceSolution + 1
  }
  return recurrenceSolution
} //O(n)

func solveHanoiTowersRecurrently(numDiscs:UInt) -> UInt {
  if numDiscs == UInt.min { return 0 }
  return 2 * solveHanoiTowersRecurrently(numDiscs-1) + 1
} //O(n)

func solveHanoiTowersClosedForm(numDiscs:Int) -> Int {
  return Int(pow(2.0, Double(numDiscs))) - 1
} //O(1)
